File: examples/chapter02/main.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"strings"

	"github.com/cloudwego/eino-ext/components/model/openai"
	"github.com/cloudwego/eino/compose"
	"github.com/cloudwego/eino/schema"
	"github.com/joho/godotenv"
)

// -- 1) Agent State
type AgentState struct {
	Order    map[string]any    `json:"order"`
	Messages []*schema.Message `json:"messages"`
}

// -- 2) Tool Implementation
func cancelOrder(orderID string) string {
	// In a real app, call backend API here
	return fmt.Sprintf("Order %s has been cancelled.", orderID)
}

func main() {
	// Load .env if present
	_ = godotenv.Load()

	ctx := context.Background()
	api_key := os.Getenv("OPENAI_API_KEY")
	base_url := os.Getenv("OPENAI_BASE_URL")

	if api_key == "" {
		fmt.Println("Error: OPENAI_API_KEY is not set")
		return
	}

	// -- 3) Initialize Model
	// We use temperature 0 for deterministic tool usage
	// temp := float32(0.0)
	chatModel, err := openai.NewChatModel(ctx, &openai.ChatModelConfig{
		APIKey:  api_key,
		BaseURL: base_url,
		Model:   "gpt-5",
		// Temperature: &temp, // Model has beta limitations, must be 1 (default)
	})
	if err != nil {
		log.Fatalf("Failed to create chat model: %v", err)
	}

	// -- 4) Define Tools
	// Schema for cancel_order
	cancelOrderTool := &schema.ToolInfo{
		Name: "cancel_order",
		Desc: "Cancel an order that hasn't shipped.",
		ParamsOneOf: schema.NewParamsOneOfByParams(map[string]*schema.ParameterInfo{
			"order_id": {
				Type:     schema.String,
				Desc:     "The ID of the order to cancel",
				Required: true,
			},
		}),
	}

	// Bind tool to model
	if err := chatModel.BindTools([]*schema.ToolInfo{cancelOrderTool}); err != nil {
		log.Fatalf("Failed to bind tools: %v", err)
	}

	// -- 5) Define Graph Nodes

	// Node: Assistant (Calls LLM)
	assistant := compose.InvokableLambda(func(ctx context.Context, state *AgentState) (*AgentState, error) {
		// Prepare System Prompt
		// In Eino, we can prepend a system message or rely on the state having one.
		// Use a dynamic system prompt based on order details if needed.
		orderID := "UNKNOWN"
		if oid, ok := state.Order["order_id"].(string); ok {
			orderID = oid
		}

		systemPrompt := fmt.Sprintf(
			"You are an e-commerce customer service agent.\n"+
				"Order ID: %s\n"+
				"If the customer asks to cancel the order, call cancel_order(order_id).\n"+
				"Then send a simple confirmation.\n"+
				"Otherwise, reply normally.",
			orderID,
		)

		// Construct messages for the LLM: System Prompt + History
		messages := []*schema.Message{schema.SystemMessage(systemPrompt)}
		messages = append(messages, state.Messages...)

		// Generate response
		resp, err := chatModel.Generate(ctx, messages)
		if err != nil {
			return nil, err
		}

		// Update state with new message
		state.Messages = append(state.Messages, resp)
		return state, nil
	})

	// Node: Tools (Executes Tools)
	toolsNode := compose.InvokableLambda(func(ctx context.Context, state *AgentState) (*AgentState, error) {
		lastMsg := state.Messages[len(state.Messages)-1]
		if len(lastMsg.ToolCalls) == 0 {
			return state, nil
		}

		for _, tc := range lastMsg.ToolCalls {
			if tc.Function.Name == "cancel_order" {
				// Parse arguments
				var args struct {
					OrderID string `json:"order_id"`
				}
				if err := json.Unmarshal([]byte(tc.Function.Arguments), &args); err != nil {
					return nil, fmt.Errorf("failed to parse args: %v", err)
				}

				// Execute Tool
				result := cancelOrder(args.OrderID)

				// Append Tool Output Message
				state.Messages = append(state.Messages, &schema.Message{
					Role:       schema.Tool,
					Content:    result,
					ToolCallID: tc.ID,
				})
			}
		}
		return state, nil
	})

	// -- 6) Construct Graph
	graph := compose.NewGraph[*AgentState, *AgentState]()

	_ = graph.AddLambdaNode("assistant", assistant)
	_ = graph.AddLambdaNode("tools", toolsNode)

	_ = graph.AddEdge(compose.START, "assistant")

	// Helper to check if we should go to tools
	shouldCallTool := func(_ context.Context, state *AgentState) (string, error) {
		lastMsg := state.Messages[len(state.Messages)-1]
		if len(lastMsg.ToolCalls) > 0 {
			return "tools", nil
		}
		return compose.END, nil
	}

	branch := compose.NewGraphBranch(shouldCallTool, map[string]bool{
		"tools":     true,
		compose.END: true,
	})

	_ = graph.AddBranch("assistant", branch)
	_ = graph.AddEdge("tools", "assistant") // Loop back after tool execution

	// Compile
	runnable, err := graph.Compile(ctx)
	if err != nil {
		log.Fatalf("Failed to compile graph: %v", err)
	}

	// -- 7) Run it
	initialOrder := map[string]interface{}{
		"order_id": "A12345",
	}
	initialMessages := []*schema.Message{
		schema.UserMessage("Please cancel my order A12345."),
	}

	initialState := &AgentState{
		Order:    initialOrder,
		Messages: initialMessages,
	}

	fmt.Println("Running Eino Agent...")
	finalState, err := runnable.Invoke(ctx, initialState)
	if err != nil {
		log.Fatalf("Agent run failed: %v", err)
	}

	// Print Results
	for _, msg := range finalState.Messages {
		fmt.Printf("[%s]: %s\n", msg.Role, msg.Content)
	}

	// -- 8) Minimal Evaluation Check
	evalOrder := map[string]interface{}{
		"order_id": "B73973",
	}
	evalMessages := []*schema.Message{
		schema.UserMessage(`Please cancel order #B73973. 
        I found a cheaper option elsewhere.`),
	}

	evalState := &AgentState{
		Order:    evalOrder,
		Messages: evalMessages,
	}

	fmt.Println("\nRunning Minimal Evaluation Check...")
	evalResult, err := runnable.Invoke(ctx, evalState)
	if err != nil {
		log.Fatalf("Eval run failed: %v", err)
	}

	// Check 1: Tool called?
	hasToolCall := false
	for _, msg := range evalResult.Messages {
		for _, tc := range msg.ToolCalls {
			if tc.Function.Name == "cancel_order" {
				hasToolCall = true
				break
			}
		}
	}
	if !hasToolCall {
		log.Fatal("Eval Failed: Cancel order tool not called")
	}

	// Check 2: Confirmation message?
	hasConfirmation := false
	for _, msg := range evalResult.Messages {
		contentLower := strings.ToLower(msg.Content)
		if strings.Contains(contentLower, "cancel") || strings.Contains(msg.Content, "取消") {
			hasConfirmation = true
			break
		}
	}
	if !hasConfirmation {
		log.Fatal("Eval Failed: Confirmation message missing")
	}

	fmt.Println("✅ Agent passed minimal evaluation.")
}

File: examples/chapter02/main.py
from dotenv import load_dotenv
load_dotenv()

from typing import Annotated, TypedDict, List
from langchain.tools import tool
from langchain_openai import ChatOpenAI
from langchain_core.messages import BaseMessage, SystemMessage, HumanMessage
from langgraph.graph import StateGraph, START, END
from langgraph.graph.message import add_messages
from langgraph.prebuilt import ToolNode, tools_condition

# -- 1) 定义我们唯一的业务工具
@tool
def cancel_order(order_id: str) -> str:
    """取消尚未发货的订单。"""
    # (在这里调用真实的后台 API)
    return f"订单 {order_id} 已被取消。"

tools = [cancel_order]

# -- 2) 定义 State
class AgentState(TypedDict):
    # add_messages 会自动处理消息的追加，而不是覆盖
    messages: Annotated[List[BaseMessage], add_messages]
    order: dict

# -- 3) 定义节点
def call_model(state: AgentState):
    order = state.get("order", {"order_id": "UNKNOWN"})
    
    # 系统提示词告诉模型具体要做什么
    system_prompt = (
        f'''你是一名电商客服代理。
        订单 ID: {order['order_id']}
        如果客户要求取消订单，请调用 cancel_order(order_id)
        然后发送一个简单的确认。
        否则，只需正常回复。'''
    )
    
    # 将系统消息放在消息列表的最前面
    # 注意：我们不在 state 中持久化 SystemMessage，而是每次调用模型时动态添加
    # 这样可以保持 state['messages'] 干净，且允许根据 order 动态改变 prompt
    messages = [SystemMessage(content=system_prompt)] + state["messages"]
    
    # 初始化模型并绑定工具
    # 使用 gpt-4o 或其他支持工具调用的现代模型
    llm = ChatOpenAI(model="gpt-5", temperature=0)
    llm_with_tools = llm.bind_tools(tools)
    
    response = llm_with_tools.invoke(messages)
    
    # 返回的内容会被 add_messages reducer 追加到 state 的 messages 中
    return {"messages": [response]}

# -- 4) 构建图
def construct_graph():
    workflow = StateGraph(AgentState)
    
    # 添加节点
    workflow.add_node("assistant", call_model)
    workflow.add_node("tools", ToolNode(tools))
    
    # 设置入口点
    workflow.add_edge(START, "assistant")
    
    # 添加条件边：
    # 如果 assistant 返回了 tool_calls，tools_condition 会路由到 "tools"
    # 否则路由到 END
    workflow.add_conditional_edges(
        "assistant",
        tools_condition,
    )
    
    # 工具执行完后，返回给 assistant 继续生成回复 (ReAct 循环)
    workflow.add_edge("tools", "assistant")
    
    return workflow.compile()

graph = construct_graph()

if __name__ == "__main__":
    example_order = {"order_id": "A12345"}
    # 注意：这里我们只传入初始的 HumanMessage
    convo = [HumanMessage(content="Please cancel my order A12345.")]
    
    # 运行图
    result = graph.invoke({"order": example_order, "messages": convo})
    
    # 打印结果
    for msg in result["messages"]:
        print(f"{msg.type}: {msg.content}")


    # 最小化评估检查
    example_order = {"order_id": "B73973"}
    convo = [HumanMessage(content='''Please cancel order #B73973. 
        I found a cheaper option elsewhere.''')]

    result = graph.invoke({"order": example_order, "messages": convo})

    # check if tool calls were made
    has_tool_call = False
    for m in result["messages"]:
        if getattr(m, "tool_calls", None):
            for tc in m.tool_calls:
                if tc["name"] == "cancel_order":
                    has_tool_call = True
                    break
    
    assert has_tool_call, "未调用取消订单工具 (Cancel order tool not called)"
        
    assert any("cancel" in m.content.lower() or "取消" in m.content for m in result["messages"]), \
        "缺失确认消息 (Confirmation message missing)"

    print("✅ Agent passed minimal evaluation.")
